metrics-publishers/cloudwatch-publisher/src/main/java/software/amazon/awssdk/metrics/publishers/cloudwatch/CloudWatchPublisher.java

```diff
@@ -0,0 +1,262 @@
+        this.publishFrequency = Validate.notNull(builder.publishFrequency, "Publish frequency cannot be null.");
+        this.namespace = Validate.notEmpty(builder.namespace, "Namespace cannot be null or empty.");
+        this.metricQueueSize = Validate.isPositive(builder.metricQueueSize, "Metric queue size should be positive.");
+        this.queue = new LinkedBlockingQueue<>(this.metricQueueSize);
+
+        this.producer = Producer.builder().queue(queue).build();
+        this.producerExecutorService = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+                                                                             .threadNamePrefix("sdk-metrics-cw-producer")
+                                                                             .build());
+
+        this.consumer = Consumer.builder()
+                               .cloudWatchClient(client)
+                               .queue(queue)
+                               .namespace(namespace)
+                               .build();
+        this.consumerExecutorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactoryBuilder()
+                                                                                      .threadNamePrefix("sdk-metrics-cw-consumer")
+                                                                                      .build());
+    }
+
+    /**
+     * CloudWatch publisher converts the given metrics into {@link MetricDatum} instances and add them
+     * to a queue for publishing.
+     *
+     * @param metricsRegistry registry containing the collected metrics
+     */
+    @Override
+    public void registerMetrics(MetricRegistry metricsRegistry) {
+        try {
+            if (publishStarted.compareAndSet(false, true)) {
+                consumerExecutorService.scheduleAtFixedRate(this::publish,
+                                                            0L,
+                                                            publishFrequency.toMillis(),
+                                                            TimeUnit.MILLISECONDS);
+            }
+
+            producerExecutorService.execute(() -> producer.addMetrics(metricsRegistry));
+        } catch (Throwable throwable) {
+            log.debug(() -> "An error occurred when registering metrics in the publisher", throwable);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> publish() {
+        List<CompletableFuture<PutMetricDataResponse>> futures = new ArrayList<>();
+        int count = 0;
+
+        try {
+            while (queue.peek() != null && ++count <= MAX_SERVICE_CALLS_PER_PUBLISH) {
+                futures.add(consumer.call());
+            }
+        } catch (Throwable throwable) {
+            log.debug(() -> "An error occurred when uploading metrics to CloudWatch.", throwable);
+        }
+
+        CompletableFuture<Void> finalFuture = CompletableFuture.allOf(futures.toArray(new CompletableFuture[futures.size()]));
```

Comment:
Do we need to cancel other futures if one of future fails of non-recoverable errors like 400? because it's likely all the futures will fail of the same reason. We are doing it for transfermanager.

https://github.com/aws/aws-sdk-java-v2/blob/s3-transfermanager/utils/src/main/java/software/amazon/awssdk/utils/CompletableFutureUtils.java#L113

Reply:
Why do we think that one 400 will make all 400?

Reply:
It might not necessary be 400, maybe something like permission issue and all requests would fail if the credential is not correct. We don't have to do it right now, just something to think about.